ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+				let mut new_state = last_checkpoint_state.clone();
+				for item in chain {
+					new_state.apply(item, false)?;
+				}
+				new_state.calc_next_timestamp(header, self.period);
+				block_state_by_hash.insert(header.hash(), new_state.clone());
+
+				let elapsed = backfill_start.elapsed();
+				info!(target: "engine",
+						"Back-filling succeed, took {} ms.",
+						elapsed.as_secs() * 1000 + elapsed.subsec_millis() as u64,
+				);
+
+				Ok(new_state)
+			}
+		}
+	}
+}
+
+impl Engine<EthereumMachine> for Clique {
+	fn name(&self) -> &str { "Clique" }
+
+	fn machine(&self) -> &EthereumMachine { &self.machine }
+	/// Clique use same fields, nonce + mixHash
+	fn seal_fields(&self, _header: &Header) -> usize { 2 }
+	fn maximum_uncle_count(&self, _block: BlockNumber) -> usize { 0 }
+
+	// No Uncle in Clique
+	fn maximum_uncle_age(&self) -> usize { 0 }
+
+	fn on_new_block(
+		&self,
+		_block: &mut ExecutedBlock,
+		_epoch_begin: bool,
+		_ancestry: &mut Iterator<Item=ExtendedHeader>,
+	) -> Result<(), Error> {
+		Ok(())
+	}
+
+	fn on_close_block(&self, _block: &mut ExecutedBlock) -> Result<(), Error> {
+		// Clique has no block reward.
+		Ok(())
+	}
+
+	fn on_seal_block(&self, block: &ExecutedBlock) -> Result<Option<Header>, Error> {
+		trace!(target: "engine", "on_seal_block");
+
+		let mut header = block.header().clone();
+
+		let state = self.state_no_backfill(header.parent_hash()).ok_or_else(
+			|| format!("on_seal_block: Unable to get parent state: {}", header.parent_hash())
+		)?;
+
+		let is_checkpoint = header.number() % self.epoch_length == 0;
+
+		header.set_author(NULL_AUTHOR);
+
+		// cast an random Vote if not checkpoint
+		if !is_checkpoint {
+			let votes = self.proposals.read().iter()
+				.filter(|(address, vote_type)| state.is_valid_vote(*address, **vote_type))
+				.map(|(address, vote_type)| (*address, *vote_type))
+				.collect_vec();
+
+			if !votes.is_empty() {
+				// Pick an random vote.
+				let random_vote = rand::thread_rng().gen_range(0 as usize, votes.len());
+				let (beneficiary, vote_type) = votes[random_vote];
+
+				trace!(target: "engine", "Casting vote: beneficiary {}, type {:?} ", beneficiary, vote_type);
+
+				header.set_author(beneficiary);
+
+				header.set_seal(
+					match vote_type {
+						VoteType::Add => { vec!(encode(&NULL_MIXHASH.to_vec()), encode(&NONCE_AUTH_VOTE.to_vec())) }
+						VoteType::Remove => { vec!(encode(&NULL_MIXHASH.to_vec()), encode(&NONCE_DROP_VOTE.to_vec())) }
+					}
+				)
+			}
+		}
+
+		// Work on clique seal.
+
+		let mut seal: Vec<u8> = Vec::with_capacity(SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH);
+
+		// At this point, extra_data should only contain miner vanity.
+		if header.extra_data().len() > SIGNER_VANITY_LENGTH {
+			warn!(target: "engine", "on_seal_block: unexpected extra extra_data: {:?}", header);
+		}
+		// vanity
+		{
+			let mut vanity = header.extra_data()[0..SIGNER_VANITY_LENGTH - 1].to_vec();
+			vanity.resize(SIGNER_VANITY_LENGTH, 0u8);
```

Comment:
this quite odd IMO because vanity will always be of size `SIGNER_VANITY_LENGTH - 1` why not just `vanity.push(0_u8)`?